frame/offences/benchmarking/src/lib.rs

```diff
@@ -0,0 +1,167 @@
+
+use pallet_im_online::{Trait as ImOnlineTrait, UnresponsivenessOffence};
+use pallet_offences::{Trait as OffencesTrait, Module as OffencesModule};
+use pallet_staking::{
+	Module as Staking, Trait as StakingTrait, RewardDestination, ValidatorPrefs,
+	Exposure, IndividualExposure, ElectionStatus
+};
+use pallet_session::Trait as SessionTrait;
+use pallet_session::historical::{Trait as HistoricalTrait, IdentificationTuple};
+
+const SEED: u32 = 0;
+
+
+const MAX_USERS: u32 = 1000;
+const MAX_REPORTERS: u32 = 100;
+const MAX_OFFENDERS: u32 = 100;
+const MAX_NOMINATORS: u32 = 100;
+
+pub struct Module<T: Trait>(OffencesModule<T>);
+
+pub trait Trait: SessionTrait + StakingTrait + OffencesTrait + ImOnlineTrait {}
+
+fn create_offender<T: Trait>(n: u32, nominators: u32) -> Result<T::AccountId, &'static str> {
+	let stash: T::AccountId = account("stash", n, SEED);
+	let controller: T::AccountId = account("controller", n, SEED);
+	let controller_lookup: <T::Lookup as StaticLookup>::Source = T::Lookup::unlookup(controller.clone());
+	let reward_destination = RewardDestination::Staked;
+	let amount = T::Currency::minimum_balance();
+
+	Staking::<T>::bond(
+		RawOrigin::Signed(stash.clone()).into(),
+		controller_lookup.clone(),
+		amount.clone(),
+		reward_destination.clone(),
+	)?;
+
+	let validator_prefs = ValidatorPrefs {
+		commission: Perbill::from_percent(50),
+	};
+	Staking::<T>::validate(RawOrigin::Signed(controller.clone()).into(), validator_prefs)?;
+
+	let mut individual_exposures = vec![];
+
+	// Create n nominators
+	for i in 0 .. nominators {
+		let nominator_stash: T::AccountId = account("nominator stash", n * MAX_NOMINATORS + i, SEED);
+		let nominator_controller: T::AccountId = account("nominator controller", n * MAX_NOMINATORS + i, SEED);
+		let nominator_controller_lookup: <T::Lookup as StaticLookup>::Source = T::Lookup::unlookup(nominator_controller.clone());
+
+		Staking::<T>::bond(
+			RawOrigin::Signed(nominator_stash.clone()).into(),
+			nominator_controller_lookup.clone(),
+			amount,
+			reward_destination,
+		)?;
+
+		let selected_validators: Vec<<T::Lookup as StaticLookup>::Source> = vec![controller_lookup.clone()];
+		Staking::<T>::nominate(RawOrigin::Signed(nominator_controller.clone()).into(), selected_validators)?;
+
+		individual_exposures.push(IndividualExposure {
+			who: nominator_controller.clone(),
+			value: amount.clone(),
+		});
+	}
+
+	let exposure = Exposure {
+		total: amount.clone() * n.into(),
+		own: amount,
+		others: individual_exposures,
+	};
+	let current_era = 0u32;
+	Staking::<T>::add_era_stakers(current_era.into(), stash.clone().into(), exposure);
+
+	return Ok(controller)
+}
+
+fn make_inputs<T: Trait>(r: u32, o: u32, n: u32)
+	-> Result<(Vec<T::AccountId>, UnresponsivenessOffence<IdentificationTuple<T>>), &'static str> {
+
+	// Make reporters.
+	let mut reporters = vec![];
+	for i in 0 .. r {
+		let reporter = account("reporter", i, SEED);
+		reporters.push(reporter);
+	}
```

Comment:
```suggestion
	let reporters = (0..r).map(|i| account("reporter", i, SEED)).collect::<Vec<T::AccountId>>();
```

Reply:
Somehow I'm starting to prefer the non-functional style. It's ok in one-liners. But sometimes people end up writing  layers of functional style, and imo it obscures the logic.

Reply:
Well, I'm myself guilty in this pr.

Reply:
I agree it is personal. discard if you prefer to do so.

Reply:
Although for the sake of completeness creating a `let a = vec![]` and push is the least performant since it requires numerous re-allocation. I think both collecting an iterator and `::with_capacity` would be better. And indeed none of it matters in this context.